packages/datadog-plugin-rhea/src/index.js

```diff
@@ -1,73 +1,92 @@
+let popIf
+
+function patchCircularBuffer (CircularBuffer) {
+  popIf = CircularBuffer.prototype.pop_if
+  CircularBuffer.prototype.pop_if = function (fn) {
+    const wrappedFn = entry => {
+      const shouldPop = fn(entry)
+      if (shouldPop && entry._dd) {
+        const state = entry.remote_state ? entry.remote_state.constructor.composite_type : 'accepted'
+        finish(entry, state)
+      }
+      return shouldPop
+    }
+    return popIf.call(this, wrappedFn)
+  }
+}
+
 function createWrapSend (tracer, config) {
   return function wrapSend (send) {
     return function sendWithTrace (msg, tag, format) {
-      const name = this.options.target && this.options.target.address
+      const name = this.options && this.options.target && this.options.target.address
         ? this.options.target.address : 'amq.topic'
-      return tracer.trace('rhea.sender.send', { tags: {
-        'resource.name': name,
-        'service.name': config.service || `${tracer._service}-amqp`,
-        'span.kind': 'producer',
-        'amqp.link.target.address': name,
-        'amqp.link.role': 'sender',
-        'out.host': this.connection.options.host,
-        'out.port': this.connection.options.port
-      } }, (span, done) => {
-        msg.delivery_annotations = msg.delivery_annotations || {}
-        tracer.inject(span, 'text_map', msg.delivery_annotations)
-        const delivery = send.apply(this, arguments)
-        if (this.options.snd_settle_mode !== 1) {
-          delivery._dd = { done, span }
-        } else {
-          done()
+      let host
+      let port
+      if (this.connection && this.connection.options) {
+        if (this.connection.options.host !== undefined) { host = this.connection.options.host }
+        if (this.connection.options.port !== undefined) { port = this.connection.options.port }
+      }
+      if (!CircularBuffer && this.session && this.session.outgoing && this.session.outgoing.deliveries) {
+        CircularBuffer = this.session.outgoing.deliveries.constructor
+        patchCircularBuffer(CircularBuffer)
+      }
+      return tracer.trace('amqp.send', {
+        tags: {
+          'component': 'rhea',
+          'resource.name': name,
+          'service.name': config.service || `${tracer._service}-amqp-producer`,
+          'span.kind': 'producer',
+          'amqp.link.target.address': name,
+          'amqp.link.role': 'sender',
+          'out.host': host,
+          'out.port': port
+        }
+      }, (span, done) => {
+        if (msg) {
+          msg.delivery_annotations = msg.delivery_annotations || {}
+          tracer.inject(span, 'text_map', msg.delivery_annotations)
         }
+        const delivery = send.apply(this, arguments)
+        delivery._dd = { done, span }
         return delivery
       })
     }
   }
 }
 
 function createWrapConnectionDispatch (tracer, config) {
+  function addTags (entry, error) {
+    if (entry && entry._dd) {
+      const { span, done } = entry._dd
+      if (error) {
```

Comment:
Unnecessary check that is already done in `addTags`.

Reply:
Where? Line 74 may be result in an `undefined`.

Reply:
`span.addTags()` will ignore tags where an undefined value is passed, so the explicit check for `error` is not needed.